cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/SwitchesChain.java

```diff
@@ -31,232 +29,173 @@
-        if (terminalEnd1 != null && terminalEnd1.getConnectable().getType() == IdentifiableType.DANGLING_LINE) {
-            return terminalEnd1;
-        }
-        if (terminalEnd2 != null && terminalEnd2.getConnectable().getType() == IdentifiableType.DANGLING_LINE) {
-            return terminalEnd2;
+        List<Integer> nodes = expandNodeBreaker(vl, end, sw);
+        if (nodes.contains(otherEnd)) {
+            return null;
         }
-        return null;
+        return uniqueTerminalAssociatedConductingEquipmentNodeBreaker(vl, nodes);
     }
 
-    private static class NodeConnectables {
-
-        private final Map<String, List<Connectable<?>>> connectables;
-
-        NodeConnectables(VoltageLevel vl, boolean isNodeBreakerModel) {
-            connectables = new HashMap<>();
+    private Terminal findTerminalChainEndBusBranch(String end, String otherEnd) {
 
-            if (isNodeBreakerModel) {
-                connectablesNodeBreaker(vl, connectables);
-            } else {
-                connectablesBusBreaker(vl, connectables);
-            }
+        List<String> configuredBuses = expandBusBranch(vl, end, sw);
+        if (configuredBuses.contains(otherEnd)) {
+            return null;
         }
+        return uniqueTerminalAssociatedConductingEquipmentBusBranch(vl, configuredBuses);
+    }
 
-        private static void connectablesNodeBreaker(VoltageLevel vl, Map<String, List<Connectable<?>>> connectables) {
-            vl.getConnectables().forEach(c -> {
-                if (isDiscarded(c)) {
-                    return;
-                }
-                c.getTerminals().stream().filter(t -> isSameVoltageLevel((Terminal) t, vl)).forEach(
-                    t -> connectables.computeIfAbsent(String.valueOf(((Terminal) t).getNodeBreakerView().getNode()),
-                        k -> new ArrayList<>()).add(c));
-            });
-        }
+    private static List<Integer> expandNodeBreaker(VoltageLevel voltageLevel, int node, Switch swTerminal) {
+        List<Integer> nodes = new ArrayList<>();
+        nodes.add(node);
 
-        private static void connectablesBusBreaker(VoltageLevel vl, Map<String, List<Connectable<?>>> connectables) {
-            vl.getConnectables().forEach(c -> {
-                if (isDiscarded(c)) {
-                    return;
-                }
-                c.getTerminals()
-                    .forEach(t -> {
-                        if (!((Terminal) t).getVoltageLevel().equals(vl)) {
-                            return;
-                        }
-                        connectables.computeIfAbsent(((Terminal) t).getBusBreakerView().getBus().getId(), k -> new ArrayList<>()).add(c);
-                    });
-            });
-        }
+        VoltageLevel.NodeBreakerView.TopologyTraverser traverser = (node1, sw, node2) -> {
+            if (sw == swTerminal) {
+                return TraverseResult.TERMINATE_PATH;
+            }
+            nodes.add(node2);
```

Comment:
In both expand methods, what if a switch is opened? shouldn't the switch chain be with only closed switches?

Reply:
We have to consider both types of switches (opened and closed). The new terminal associated to the TieFlow should always have the same flow as the initial terminal defined in the switch, independently of the status of the switches in the chain. But this assertion is guaranteed in a chain, the active power flow will be the same at any node  independently of the status of the switches.

Reply:
Thanks for explaining! I saw you also added a comment clarifying that :+1: